### PR TITLE
Fix S3459 FP: Backing field with ref property

### LINKED ISSUE #9106: Fix S3459 FP: Backing field with `ref` property
### Description

False Positive of S3459 for backing field with `ref` property.

### Repro steps
```csharp
public class MyClass
{
    private int _foo;

    public ref int Foo => ref _foo;
}
```

### Expected behavior

No warning

### Actual behavior

warning S3459: Remove unassigned field '_foo', or set its value. (https://rules.sonarsource.com/csharp/RSPEC-3459)

### Known workarounds

Supress this warning manually

### Related information

* C#/VB.NET Plugins version: SonarAnalyzer.CSharp 9.23.2.88755
* Visual Studio version 17.9.2
* MSBuild / dotnet version: dotnet 8.0.200
* SonarScanner for .NET version (if used)
* Operating System: Windows 10

### PR DESCRIPTION
**Pull request by martin-strecker-sonarsource**
Fixes #9106

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CSharp/Rules/NotAssignedPrivateMember.cs [hunk 1]
```diff
@@ -184,7 +184,8 @@ private static ISet<ISymbol> GetAssignedMemberSymbols(IList<MemberUsage> memberU
 
                 if (PreOrPostfixOpSyntaxKinds.Contains(parentNode.Kind())
                     || (parentNode is AssignmentExpressionSyntax assignment && assignment.Left == node)
-                    || (parentNode is ArgumentSyntax argument && (!argument.RefOrOutKeyword.IsKind(SyntaxKind.None) || TupleExpressionSyntaxWrapper.IsInstance(argument.Parent))))
+                    || (parentNode is ArgumentSyntax argument && (!argument.RefOrOutKeyword.IsKind(SyntaxKind.None) || TupleExpressionSyntaxWrapper.IsInstance(argument.Parent)))
+                    || RefExpressionSyntaxWrapper.IsInstance(parentNode))
```
Comment (Tim-Pohlmann) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/NotAssignedPrivateMember.cs:188]: This is too simple and introduces an FN (run with C# 9):
```
    private int _foo; // FN
    public int Foo => _foo;
    void Method()
    {
        _ = ref _foo;
    }
```

File: analyzers/src/SonarAnalyzer.CSharp/Rules/NotAssignedPrivateMember.cs
Comment (Tim-Pohlmann) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/NotAssignedPrivateMember.cs:1]: Change to file-scoped namespace.

**Review by Tim-Pohlmann — CHANGES_REQUESTED** [review]
Round 1 (many more to come!)

File: analyzers/src/SonarAnalyzer.CSharp/Rules/NotAssignedPrivateMember.cs (hunk 1 above)
Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/NotAssignedPrivateMember.cs:188]: This is not a FN. If you get a reference to the field, you have write access to it and that write can happen anywhere. This is like calling `SomeMethod(ref field)` where you do not know what happens inside `SomeMethod`. We would need to do symbolic execution on the assigned variable to properly detect all scenarios.

Comment (Tim-Pohlmann) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/NotAssignedPrivateMember.cs:188]: But you have access to the method where the ref was used. I accept that this is out of the scope of this ticket, but at least add a reproducer to document the limitation.

**Review by Tim-Pohlmann — APPROVED** [review]
Nice improvements!

**Comment by CristianAmbrosini** [comment]
Peach validation
Lost: 12 issues (TN)
Added: /

**Pull request merged**